### PR TITLE
Add disambiugator to ExpnData

### PR DESCRIPTION
**Pull request by Aaron1011**
I still need to write a bunch of comments. Opening to see how bad the perf impact is.

cc https://github.com/rust-lang/rust/issues/79560

### CODE REVIEW

**Comment by Aaron1011** [comment]
I think this should actually use `HashStable` to determine when we need a disambiguation - theoretically, we could end up with two `ExpnData`s with distinct `parent` fields, but which end up getting the same `Fingerprint` when we recursively hash them. However, we don't have a `HashStableContext` available when we create an `ExpnId`, so this could be tricky.

**Comment by Aaron1011** [comment]
This has a tricky interaction with incremental compilation, and will require some additional thought.

**Comment by Aaron1011** [comment]
Blocked on https://github.com/rust-lang/rust/pull/80648

**Comment by Aaron1011** [comment]
@petrochenkov: This is now ready for review

**Comment by petrochenkov** [comment]
(I probably won't get to this until the next weekend.)

File: compiler/rustc_span/src/lib.rs [hunk 1]
```diff
@@ -1966,14 +1972,13 @@ impl<CTX: HashStableContext> HashStable<CTX> for SyntaxContext {
     }
 }
 
+pub type ExpnIdCache = RefCell<Vec<Option<Fingerprint>>>;
+
 impl<CTX: HashStableContext> HashStable<CTX> for ExpnId {
     fn hash_stable(&self, ctx: &mut CTX, hasher: &mut StableHasher) {
         // Since the same expansion context is usually referenced many
         // times, we cache a stable hash of it and hash that instead of
         // recursing every time.
```
Comment (petrochenkov) [inline compiler/rustc_span/src/lib.rs:1981]: The comment is now outdated.

File: compiler/rustc_span/src/hygiene.rs [hunk 2]
```diff
@@ -1272,3 +1305,116 @@ impl<D: Decoder> Decodable<D> for SyntaxContext {
...
+                true
+            }
+        });
+
+    if modified {
+        info!("Set disambiguator for {:?} (hash {:?})", expn_id, first_hash);
+        info!("expn_data = {:?}", expn_id.expn_data());
+
+        // Verify that the new disambiguator makes the hash unique
+        #[cfg(debug_assertions)]
```
Comment (petrochenkov) [inline compiler/rustc_span/src/hygiene.rs:1407]: Do we have debug assertions for hash collisions like this anywhere else in the compiler? This seems a bit excessive.

File: compiler/rustc_span/src/hygiene.rs [hunk 3]
```diff
@@ -1272,3 +1305,116 @@ impl<D: Decoder> Decodable<D> for SyntaxContext {
...
+/// Updates the `disambiguator` field of the corresponding `ExpnData`
+/// such that the `Fingerprint` of the `ExpnData` does not collide with
+/// any other `ExpnIds`.
+///
+/// This method is called only when an `ExpnData` is first associated
+/// with an `ExpnId` (when the `ExpnId` is initially constructed, or via
+/// `set_expn_data`). It is *not* called for foreign `ExpnId`s deserialized
+/// from another crate's metadata - since `ExpnData` includes a `krate` field,
+/// collisions are only possible between `ExpnId`s within the same crate.
+pub(super) fn update_disambiguator(expn_id: ExpnId) {
```
Comment (petrochenkov) [inline compiler/rustc_span/src/hygiene.rs:1318]:
```suggestion
fn update_disambiguator(expn_id: ExpnId) {
```

Comment (petrochenkov) [reply compiler/rustc_span/src/hygiene.rs:1318]: This also should be a method on `HygieneData` if possible.
Inside this functions we take the hygiene data lock (e.g. in `expn_id.expn_data()`), but that is extra work - at call points of `update_disambiguator` we already have to take that lock.

File: compiler/rustc_span/src/hygiene.rs [hunk 4]
```diff
@@ -1272,3 +1305,116 @@ impl<D: Decoder> Decodable<D> for SyntaxContext {
...
+        expn_data.disambiguator, 0,
+        "Already set disambiguator for ExpnData: {:?}",
+        expn_data
+    );
+    expn_data.hash_stable(&mut ctx, &mut hasher);
+    let first_hash = hasher.finish();
+
+    let modified =
+        HygieneData::with(|data| match data.expn_data_disambiguators.entry(first_hash) {
+            Entry::Vacant(e) => {
```
Comment (petrochenkov) [inline compiler/rustc_span/src/hygiene.rs:1390]: Is it possible to avoid explicit entry API here with `expn_data_disambiguators.entry(first_hash).or_default()` or other method? It's too damn ugly otherwise.

File: compiler/rustc_span/src/hygiene.rs [hunk 5]
```diff
@@ -110,8 +120,9 @@ impl ExpnId {
             let old_expn_data = &mut data.expn_data[self.0 as usize];
             assert!(old_expn_data.is_none(), "expansion data is reset for an expansion ID");
             expn_data.orig_id.replace(self.as_u32()).expect_none("orig_id should be None");
-            *old_expn_data = Some(expn_data);
-        })
+            data.expn_data[self.0 as usize] = Some(expn_data);
```
Comment (petrochenkov) [inline compiler/rustc_span/src/hygiene.rs:123]: Why was this changed?

File: compiler/rustc_span/src/hygiene.rs [hunk 6]
```diff
@@ -722,10 +740,23 @@ pub struct ExpnData {
...
+    /// implementation).
+    ///
+    /// The `disambiguator` field is set by `update_disambiguator` when two distinct
+    /// `ExpnId`s would end up with the same `Fingerprint`. Since `ExpnData` includes
+    /// a `krate` field, this value only needs to be unique within a single crate.
+    disambiguator: u32,
 }
 
-// This would require special handling of `orig_id` and `parent`
+// These would require special handling of `orig_id` and `parent`
```
Comment (petrochenkov) [inline compiler/rustc_span/src/hygiene.rs:757]:
```suggestion
// These would require special handling of `orig_id`.
```
`parent` is now hashed.

File: compiler/rustc_span/src/hygiene.rs (hunk 2 above)
Comment (Aaron1011) [reply compiler/rustc_span/src/hygiene.rs:1407]: I don't think we have this kind of problem anywhere else in the compiler (AFAIK, `DefId`s are the only other place where we generate disambiguators, and that's done before any hashing).

File: compiler/rustc_span/src/hygiene.rs (hunk 3 above)
Comment (Aaron1011) [reply compiler/rustc_span/src/hygiene.rs:1318]: We cannot call this method with the hygiene data lock taken, since hashing the `ExpnData` needs to take the lock.

**Comment by Aaron1011** [comment]
@petrochenkov I've addressed your comments

**Comment by petrochenkov** [comment]
r=me with commits squashed.
